### PR TITLE
KAS: Improve TypeScript types

### PR DESCRIPTION
**Pull request by jeremywiebe**
## Summary:

While working on splitting validation (empty checking) from scoring, I couldn't get types for `KAS` functions even though everything seemed to be in TypeScript. Well, the entry file `index.js` was still in JS and that blocked any TypeScript types from bubbling outside of the `@khanacademy/kas` package. I've renamed the entry file `index.js` => `index.ts` and that gets the types flowin' real good. I had to fix some type issues and a tiny bit of code but I think this is a big net improvement.

Issue: LEMS-2598

## Test plan:

`yarn typecheck`
`yarn test`

### CODE REVIEW

File: packages/kas/package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
     },
     "module": "dist/es/index.js",
     "main": "dist/index.js",
-    "source": "src/index.js",
+    "source": "src/index.ts",
```
Comment (jeremywiebe) [inline packages/kas/package.json:20]: 🎉

File: packages/kas/src/__tests__/transforming.test.ts [hunk 2]
```diff
@@ -9,29 +9,44 @@ expect.extend({
 
         return actual === expected
             ? {pass: true, message: () => ""}
-            : {pass: false, message: () => `${input} factors as ${reference}`};
+            : {
```
Comment (jeremywiebe) [inline packages/kas/src/__tests__/transforming.test.ts:12]: These custom assertions were stating that the input matches the expected value when they failed, which was super confusing. 

For some, I was able to use Jest's internal utils for printing the expected and actual values as well as providing a nice "over/under" diff. 

Ref: https://dev.to/zirkelc/improve-your-testing-with-custom-matchers-in-jest-2ibm

File: packages/kas/src/types.d.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export type ExpressionVars = {
     equalIgnoringCase: boolean;
 };
 
+// TODO: Convert to interface and make the various Expr's implement it?
 export type Expression = {
```
Comment (jeremywiebe) [inline packages/kas/src/types.d.ts:21]: All of the custom classes in `nodes.ts` conform to this type, but implicitly. That's not obvious.

**Review by kevinb-khan — APPROVED** [review]
LGTM.  Thanks for changing `index.js` to `index.ts` and fixing all of the TypeScript errors that resulted from that change.

**Pull request merged**